cupy/core/_fusion_analysis.py

```diff
@@ -0,0 +1,591 @@
+        return x.as_interface()
+
+    def _from_arraylike_interface(self, x):
+        """Returns ``_FusionCuda{Array/Scalar}`` object from the input.
+        """
+        if isinstance(x, _fusion_interface._FusionVariableInterfaceBase):
+            return x.content
+        if isinstance(x, _accepted_types):
+            dtype = numpy.dtype(type(x))
+            return self.vc.generate_new_scalar(dtype, const_value=x)
+        if isinstance(x, (numpy.ndarray, core.ndarray)):
+            raise TypeError('Concrete ndarray is not supported in fusion.')
+        raise TypeError('{} type is not supported'.format(type(x)))
+
+    def _from_interface(self, x):
+        """Returns ``_FusionCuda{Array/Scalar}`` object or ``None``.
+        """
+        if x is None:
+            return None
+        return self._from_arraylike_interface(x)
+
+    def call_ufunc(self, ufunc, *args, **kwargs):
+        """Register an elementwise operation with the given parameters.
+
+        Args:
+            ufunc(_kernel.ufunc): The ufunc to operate.
+            args(tuple): The arguments.
+            kwargs(dict): The keyword arguments.
+        """
+
+        assert isinstance(ufunc, _kernel.ufunc)
+
+        # Parse Inputs.
+        nin = ufunc.nin
+        nout = ufunc.nout
+        dtype = kwargs.pop('dtype', None)
+
+        if 'out' in kwargs and len(args) > nin:
+            raise ValueError(
+                'cannot specify \'out\' as both a positional and '
+                'keyword argument')
+
+        in_params = [self._from_arraylike_interface(x) for x in args[:nin]]
+        out_params = [
+            self._from_interface(x)
+            for x in args[nin:] + (kwargs.pop('out', None),)
+            if x is not None
+        ]
+        params = in_params + out_params
+
+        if len(kwargs) > 0:
+            raise TypeError('Wrong arguments {}'.format(kwargs))
+        if len(in_params) != nin or len(out_params) > nout:
+            raise ValueError('Invalid number of arguments')
+        if not all([isinstance(v, _FusionCudaArray) for v in out_params]):
+            raise TypeError('Return arrays must be of ArrayType')
+
+        # Check for inplace operation.
+        for i, out_param1 in enumerate(out_params):
+            for out_param2 in out_params[:i]:
+                if out_param1.memory == out_param2.memory:
+                    # NumPy does not raise this error.
+                    raise ValueError('Outputs of ufunc must not share memory')
+
+        for i, in_param in enumerate(in_params):
+            should_copy = any([
+                in_param.memory == out_param.memory and in_param != out_param
+                for out_param in out_params
+            ])
+            if should_copy:
+                in_params[i] = self._from_arraylike_interface(
+                    self.call_ufunc(
+                        core.elementwise_copy,
+                        self._to_interface(in_param)))
+
+        # Broadcast shapes
```

Comment:
What's the difference between "rshape" and "ashape"?
Is it possible to make some comments?